### PR TITLE
Test clean-up: Move pod-related utils to libpod

### PR DESCRIPTION
**Pull request by alromeros**
### What this PR does

This PR introduces the following changes:

- Pod-running functions such as `RunPodInNamespace` and `RunPod` have been moved from `test/utils.go` to a new file in the libpod package.
- Two minor functions (`RemoveHostDiskImage` and `RandTmpDir`) have been moved from test/utils.go to libstorage since they were just used for a couple image-related operations.

### Release note

```release-note
NONE
```

### CODE REVIEW

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
Thanks this important PR.

/sig code-quality

File: tests/libpod/run.go
Comment (EdDev) [inline tests/libpod/run.go]: This change better come in a different commit. Perhaps in the commit that brought this code in this file.

File: tests/libpod/run.go
Comment (EdDev) [inline tests/libpod/run.go:38]: It would be a good opportunity to follow up on this and:
- Just call it `Run` as it has the context of the package now. Unlike `Create`, it also waits to confirm running, that is why it is called differently.
- Require to input the namespace at the caller, do not assume it is part of the object (i.e. the current `RunPod` implementation).
- Replace asserts in the implementation with a returned error, allowing the caller to assert it and avoid the problem of ginkgo tracking the source of the error (i.e. to use of offset).

**Review by EdDev — COMMENTED** [review]
> Test cleanup: Create a new file in libpod to contain pod-running utils 

It will be nice if you could express in the topic that you treat the tests and utils file, then mention what moved. The new file in the `libpod` is an implementation detail and not the main point here.

E.g.: `tests, utils: Relocate `RunPod*` helpers to 'libpod'`

If you also agree to change the helper name, then some reasoning int he body will be needed.

File: tests/libstorage/image.go
Comment (alromeros) [reply tests/libstorage/image.go:94]: Oops, then this should be in `hostdisk` for sure. I'll move them.

**Comment by alromeros** [comment]
Just noticed this PR conflicts heavily with https://github.com/kubevirt/kubevirt/pull/12134, which has just been merged. I'm unsure if these changes are still relevant after the merge, are we ok with them? Should we still rename `pod.go` to `run.go`? 
/cc @assafad

File: tests/storage/export.go
Comment (dankenigsberg) [inline tests/storage/export.go:208]: I hope that in a future PR, the `err` can be properly returned instead of being swallowed here. It is hard to debug failed `Expect`s that are hidden inside functions.

File: tests/storage/hostdisk.go
Comment (dankenigsberg) [inline tests/storage/hostdisk.go:54]: I find the asymmetry between  `CreateDiskOnHost` and `RemoveHostDiskImage` almost painful.

Would you consider having a follow-up PR that makes them more similar e.g.
```
CreateHostDisk() (path, nodeName, error);
RemoveHostDisk(path, nodeName string) error;
```

File: tests/storage/hostdisk.go
Comment (dankenigsberg) [inline tests/storage/hostdisk.go:65]: This function should not be used for anything but generating paths for hostdisks. Can you find a better name for it (in a followup commit)? e.g `RandHostDiskDir` ?

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
thanks for the quick turnaround.

**Comment by alromeros** [comment]
Updated with a new commit following review comments

File: tests/migration/migration.go
Comment (alromeros) [reply tests/migration/migration.go]: Oops right, getting `pausePod` from the scope is a mistake on my side. I'll include the error handling in the loop. IMO this lambda could be justified but since it's only used in one loop I'm fine changing it.

File: tests/libpod/query.go
Comment (alromeros) [reply tests/libpod/query.go]: This was a minor fix to account for the removal of `getRunningPodByVirtualMachineInstance`, but I found a more transparent fix in the function that used it.

File: tests/libpod/run.go
Comment (ShellyKa13) [inline tests/libpod/run.go:46]: If you are keeping the Eventually here, maybe you should change it to be EventuallyWithOffset so the error will be shown from the context it was called from.

File: tests/migration/migration.go
Comment (ShellyKa13) [inline tests/migration/migration.go:3030]: Ill just say that the following 10 lines are pretty bad and unclear, why someone creates a list of pods just to delete them? why not instead of append do the delete? I wont ask you to change that, only if you really want lol

File: tests/storage/hostdisk.go
Comment (ShellyKa13) [inline tests/storage/hostdisk.go]: If you are moving the function its best not to add gomega to this file and have this function return error

File: tests/libpod/BUILD.bazel
Comment (ShellyKa13) [inline tests/libpod/BUILD.bazel:9]: this remove should be on the first commit as I commented there

File: tests/storage/hostdisk.go
Comment (ShellyKa13) [reply tests/storage/hostdisk.go]: I see that you did that in the last commit, I really dislike adding new commits to change stuff from previous commits.. makes anyone reviewing by commit order waste time on redundant review

File: tests/libpod/BUILD.bazel
Comment (alromeros) [reply tests/libpod/BUILD.bazel]: Updated the first commit so we remove it there.

File: tests/libpod/run.go
Comment (alromeros) [reply tests/libpod/run.go]: Makes sense, ack

File: tests/storage/hostdisk.go
Comment (alromeros) [reply tests/storage/hostdisk.go]: Well, the new commit is explicit about updating these functions to match conventions. IMO it's better to have one commit that simply moves the functions and, if necessary, other than updates them. I agree we need to avoid spaghetti commits that do things they aren't supposed to (I'm guilty of this), but I still think it's viable to have your commits reflect the development history even if they update earlier changes.

File: tests/libpod/BUILD.bazel
Comment (alromeros) [reply tests/libpod/BUILD.bazel:9]: Agree, updated

**Review by ShellyKa13 — COMMENTED** [review]
/lgtm

**Comment by alicefr** [comment]
The PR looks good to me
/approve
/hold
@EdDev can you please take another look?

File: tests/storage/hostdisk.go
Comment (EdDev) [reply tests/storage/hostdisk.go]: When reviewing per-commit like I prefer to do, we will find this issue and comment on it. Then we will find it later as fixed.
This is slowing does the reviewer and the author (which needs to read and respond).

Sometime this is just unavoidable, but sometimes the author can do this fix prior to the move as its own commit (and reasoning). It does not add more work to the reviewer, but it may add a bit more rebase work to the author.

But I also find it fine to always refer to the commit message reasoning and intention. Doing more than it says is a nice-to-have, so reviewers could take that into consideration as well.

There is no right or wrong here.

Regarding the development history: As I see it, the commits should tell a story after the fact, i.e. a non drafted story. The logic should make sense to the reader, not in the sense of how the history of the development has been composed, but how the logic explains best the story.

File: tests/storage/storage.go
Comment (EdDev) [inline tests/storage/storage.go:616]: `Expect(RemoveHostDisk(hostDiskDir, nodeName)).To(Succeed())`

File: tests/vmi_configuration_test.go
Comment (EdDev) [inline tests/vmi_configuration_test.go:2190]: Same
`Expect(...).To(Succeed())`

But this test should be fixed as a follow up as `defer` is usually not a good fit.

File: tests/vmi_configuration_test.go
Comment (EdDev) [inline tests/vmi_configuration_test.go:2390]: Same
`Expect(...).To(Succeed())`

But this test should be fixed as a follow up as defer is usually not a good fit.

**Review by EdDev — COMMENTED** [review]
Looks very nice, thanks!
I had only a few inline comments, nothing major.

>  Test cleanup: Move hostdisk manipulation functions to an appropriate package

Commit subject line is too long (see it breaks when watching on the UI).

File: tests/storage/hostdisk.go
Comment (alromeros) [reply tests/storage/hostdisk.go]: Thanks for the insight @EdDev, yeah, I especially agree with your last point.

**Comment by EdDev** [comment]
@alromeros , feel free to unhold if you are not going to change per the last nit commit.

File: tests/migration/migration.go
Comment (alromeros) [reply tests/migration/migration.go:3035]: I also don't like `break` but wanted to avoid repeating the ` pausedPod, err = libpod.Run(pausePod, ns)` line. But I can change it if preferred.

Comment (alromeros) [reply tests/migration/migration.go]: ack :+1:

**Review by EdDev — APPROVED** [review]
Looks great, thanks you!

/unhold

**Review by EdDev — APPROVED** [review]

**Pull request merged**
